### PR TITLE
Refactor integration tests with pytest

### PR DESCRIPTION
**Pull request by ThomasCluzel**
## Proposed changes

> _Describe the big picture of your changes here to communicate to the maintainers why we should accept this pull request. If it fixes a bug or resolves a feature request, be sure to link to that issue._

This pull request refactors the tests. Instead of using bash scripts, we use Python 3 with `pytest`.

## Types of changes

What types of changes does your code introduce to this project?
> _Put an `x` in the boxes that apply_

- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [x] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Documentation Update (if none of the other choices apply)

## Issues closed by changes

> _List here all issues closed by your changes. Use a list of items like `- [x] Close #0`_

## Checklist

> _Put an `x` in the boxes that apply. You can also fill these out after creating the PR. If you're unsure about any of them, don't hesitate to ask. We're here to help! This is simply a reminder of what we are going to look for before merging your code._

- [x] I have read the [CONTRIBUTING](CONTRIBUTING.md) doc
- [x] I agree with the [CODE OF CONDUCT](CODE_OF_CONDUCT.md)
- [x] Lint and unit tests pass locally with my changes
- [ ] Codacy validates the quality of the modifications
- [ ] GitHub Actions CI workflow passes
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] I have added necessary documentation (if appropriate)
- [ ] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

File: tests/test_docker_cat.py [hunk 1]
```diff
@@ -0,0 +1,539 @@
...
+        """As a user of this image, I want to analyze a fortran 77 project
+        so that I can see its level of quality on the SonarQube server.
+        """
+        sensors = (
+            "INFO: Sensor Sonar i-Code [icode]",
+        )
+        self.language("Fortran 77", "f77", "fortran77", sensors, "fortran77-dummy-project", 11)
+
+    def test_language_fortran_90(self):
+        """As a user of this image, I want to analyze a fortran 90 project
```
Comment (ThomasCluzel) [inline tests/test_docker_cat.py:403]: Codacy found an issue: [1 blank line required between summary line and description (found 0) (D205)](https://app.codacy.com/gh/cnescatlab/docker-cat/pullRequest?prid=6253839)

**Comment by ThomasCluzel** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 31
           

Complexity increasing per file
==============================
- tests/test_docker_cat.py  11
         

Clones added
============
- tests/test_docker_cat.py  2
         
```


See the complete overview on [Codacy](https://app.codacy.com/gh/cnescatlab/docker-cat/pullRequest?prid=6253839&bid=20365774)

File: tests/test_docker_cat.py [hunk 2]
```diff
@@ -0,0 +1,539 @@
...
+        sensors = (
+            "Sensor C++ (Community) SquidSensor [cxx]",
+            "Sensor SonarFrama-C [framac]"
+        )
+        self.language("C/C++", "c++", "c_cpp", sensors, "c-dummy-project", 0, "CNES_C_A", 1)
+        # 0 issue are expected with the Sonar way Quality Profile for
+        # C++ (Community) because it does not have any rule enabled.
+    
+    def test_language_fortran_77(self):
+        """As a user of this image, I want to analyze a fortran 77 project
```
Comment (ThomasCluzel) [inline tests/test_docker_cat.py:394]: Codacy found an issue: [1 blank line required between summary line and description (found 0) (D205)](https://app.codacy.com/gh/cnescatlab/docker-cat/pullRequest?prid=6253839)

File: tests/test_docker_cat.py [hunk 3]
```diff
@@ -0,0 +1,539 @@
...
+                })
+        else:
+            print(f"Using container {cls.CAT_CONTAINER_NAME}")
+        # Wait for the SonarQube server inside it to be set up
+        print(f"Waiting for {cls.CAT_CONTAINER_NAME} to be up...")
+        cls.wait_cat_ready(cls.CAT_CONTAINER_NAME)
+
+    @classmethod
+    def teardown_class(cls):
+        """Stop the container
```
Comment (ThomasCluzel) [inline tests/test_docker_cat.py:277]: Codacy found an issue: [One-line docstring should fit on one line with quotes (found 2) (D200)](https://app.codacy.com/gh/cnescatlab/docker-cat/pullRequest?prid=6253839)

File: tests/test_docker_cat.py [hunk 4]
```diff
@@ -0,0 +1,539 @@
...
+            # Get the new number of issues
+            output = requests.get(f"{cls.CAT_URL}/api/issues/search?componentKeys={project_key}",
+                auth=("admin", "admin")).json()['issues']
+            issues = [ issue for issue in output if issue['status'] in ('OPEN', 'TO_REVIEW') ]
+            # Hint: if this test fails, there should be {nb_issues_cnes_qp} issues on the {language_name} dummy project with the {cnes_qp} QP but {len(issues)} were found
+            assert len(issues) == nb_issues_cnes_qp
+
+    @classmethod
+    def analysis_tool(cls, tool: str, cmd: str, ref_file: str, tmp_file: str, store_output: bool = True):
+        """This function tests that the image can run a specified code analyzer
```
Comment (ThomasCluzel) [inline tests/test_docker_cat.py:141]: Codacy found an issue: [1 blank line required between summary line and description (found 0) (D205)](https://app.codacy.com/gh/cnescatlab/docker-cat/pullRequest?prid=6253839)

File: tests/test_docker_cat.py [hunk 5]
```diff
@@ -0,0 +1,539 @@
...
+        so that I can see its level of quality on the SonarQube server.
+        """
+        sensors = (
+            "INFO: Sensor Sonar i-Code [icode]",
+        )
+        self.language("Shell", "shell", "shell", sensors, "shell-dummy-project", 58)
+
+    # Test analysis tools
+    def test_tool_cppcheck(self):
+        """As a user of this image, I want to run cppcheck from within a container
```
Comment (ThomasCluzel) [inline tests/test_docker_cat.py:440]: Codacy found an issue: [1 blank line required between summary line and description (found 0) (D205)](https://app.codacy.com/gh/cnescatlab/docker-cat/pullRequest?prid=6253839)

File: tests/test_docker_cat.py [hunk 6]
```diff
@@ -0,0 +1,539 @@
...
+Run the tests by launching ``pytest`` from the "tests/" folder.
+
+Pytest documentation: https://docs.pytest.org/en/stable/contents.html
+"""
+
+import filecmp
+import os
+import platform
+import re
+import subprocess
```
Comment (ThomasCluzel) [inline tests/test_docker_cat.py:12]: Codacy found an issue: [Consider possible security implications associated with subprocess module.](https://app.codacy.com/gh/cnescatlab/docker-cat/pullRequest?prid=6253839)

File: tests/test_docker_cat.py [hunk 7]
```diff
@@ -0,0 +1,539 @@
...
+        Quality Gate configured and set as default so that I can use it.
+        """
+        quality_gates = requests.get(f"{self.CAT_URL}/api/qualitygates/list").json()['qualitygates']
+        cnes_quality_gates = [ gate for gate in quality_gates if gate['name'] == "CNES" ]
+        # Hint: if one of these tests fails, the CNES Quality Gate may not have been added correctly, check the container logs
+        assert cnes_quality_gates # not empty
+        assert cnes_quality_gates[0]['isDefault']
+
+    def test_check_qp(self):
+        """As a SonarQube user, I want the SonarQube server to have the
```
Comment (ThomasCluzel) [inline tests/test_docker_cat.py:349]: Codacy found an issue: [1 blank line required between summary line and description (found 0) (D205)](https://app.codacy.com/gh/cnescatlab/docker-cat/pullRequest?prid=6253839)

File: tests/test_docker_cat.py [hunk 8]
```diff
@@ -0,0 +1,539 @@
...
+            requests.post(f"{cls.CAT_URL}/api/qualityprofiles/deactivate_rule",
+                auth=("admin", "admin"),
+                data={
+                    "key": qp_key,
+                    "rule": rule_violated
+                })
+
+    @classmethod
+    def setup_class(cls):
+        """Set up the tests
```
Comment (ThomasCluzel) [inline tests/test_docker_cat.py:253]: Codacy found an issue: [1 blank line required between summary line and description (found 0) (D205)](https://app.codacy.com/gh/cnescatlab/docker-cat/pullRequest?prid=6253839)

File: tests/test_docker_cat.py [hunk 9]
```diff
@@ -0,0 +1,539 @@
...
+        sensors = (
+            "INFO: Sensor CheckstyleSensor [checkstyle]",
+            "INFO: Sensor FindBugs Sensor [findbugs]",
+            "INFO: Sensor PmdSensor [pmd]",
+            "INFO: Sensor CoberturaSensor [cobertura]"
+        )
+        self.language("Java", "java", "java", sensors, "java-dummy-project", 3, "CNES_JAVA_A", 6)
+
+    def test_language_python(self):
+        """As a user of this image, I want to analyze a Python project
```
Comment (ThomasCluzel) [inline tests/test_docker_cat.py:424]: Codacy found an issue: [1 blank line required between summary line and description (found 0) (D205)](https://app.codacy.com/gh/cnescatlab/docker-cat/pullRequest?prid=6253839)

File: tests/test_docker_cat.py [hunk 10]
```diff
@@ -0,0 +1,539 @@
...
+        so that I can see its level of quality on the SonarQube server.
+        """
+        self.language("Python", "py", "python", (), "python-dummy-project", 2, "CNES_PYTHON_A", 3)
+
+    def test_language_shell(self):
+        """As a user of this image, I want to analyze a shell project
```
Comment (ThomasCluzel) [inline tests/test_docker_cat.py:430]: Codacy found an issue: [1 blank line required between summary line and description (found 0) (D205)](https://app.codacy.com/gh/cnescatlab/docker-cat/pullRequest?prid=6253839)

File: tests/test_docker_cat.py [hunk 11]
```diff
@@ -0,0 +1,539 @@
...
+import subprocess
+import time
+from pathlib import Path
+
+import docker
+import requests
+
+
+class TestDockerCAT:
+    """This class test the lequal/docker-cat image.
```
Comment (ThomasCluzel) [inline tests/test_docker_cat.py:21]: Codacy found an issue: [1 blank line required before class docstring (found 0) (D203)](https://app.codacy.com/gh/cnescatlab/docker-cat/pullRequest?prid=6253839)

File: tests/test_docker_cat.py [hunk 12]
```diff
@@ -0,0 +1,539 @@
...
+
+    def test_tool_shellcheck(self):
+        """As a user of this image, I want to run shellcheck from within a container
+        so that it produces a report.
+        """
+        cmd = "bash -c 'shellcheck -s sh -f checkstyle tests/shell/src/script.sh || true'"
+        self.analysis_tool("shellcheck", cmd, "tests/shell/reference-shellcheck-results.xml", "tests/shell/tmp-shellcheck-results.xml")
+
+    def test_tool_vera(self):
+        """As a user of this image, I want to run Vera++ from within a container
```
Comment (ThomasCluzel) [inline tests/test_docker_cat.py:482]: Codacy found an issue: [1 blank line required between summary line and description (found 0) (D205)](https://app.codacy.com/gh/cnescatlab/docker-cat/pullRequest?prid=6253839)

File: tests/test_docker_cat.py [hunk 13]
```diff
@@ -0,0 +1,539 @@
...
+        )
+        quality_profiles = requests.get(f"{self.CAT_URL}/api/qualityprofiles/search").json()['profiles']
+        cnes_quality_profiles = [ qp['name'] for qp in quality_profiles if re.match(r'CNES_\w+_[ABCD]', qp['name']) ]
+        for profile in required_quality_profiles:
+            # Hint: if this test fails, one or more Quality Profiles may be missing, check the container logs
+            assert profile in cnes_quality_profiles
+
+    # Language tests
+    def test_language_c_cpp(self):
+        """As a user of this image, I want to analyze a C/C++ project
```
Comment (ThomasCluzel) [inline tests/test_docker_cat.py:382]: Codacy found an issue: [1 blank line required between summary line and description (found 0) (D205)](https://app.codacy.com/gh/cnescatlab/docker-cat/pullRequest?prid=6253839)

File: tests/test_docker_cat.py [hunk 14]
```diff
@@ -0,0 +1,539 @@
+"""Automated integration test of Docker CAT
+
+Run the tests by launching ``pytest`` from the "tests/" folder.
+
+Pytest documentation: https://docs.pytest.org/en/stable/contents.html
+"""
+
+import filecmp
+import os
+import platform
```
Comment (ThomasCluzel) [inline tests/test_docker_cat.py:10]: Codacy found an issue: [Unused import platform](https://app.codacy.com/gh/cnescatlab/docker-cat/pullRequest?prid=6253839)

File: tests/test_docker_cat.py [hunk 15]
```diff
@@ -0,0 +1,539 @@
...
+        so that it produces a report.
+        """
+        ref = "tests/c_cpp/reference-vera-results.xml"
+        output = "tests/c_cpp/tmp-vera-results.xml"
+        cmd = f"vera++ -s -c {output} tests/c_cpp/vera/main.cpp"
+        self.analysis_tool("Vera++", cmd, ref, output, False)
+
+    # Test importation of analysis results
+    def test_import_cppcheck_results(self):
+        """As a user of this image, I want to be able to import the results
```
Comment (ThomasCluzel) [inline tests/test_docker_cat.py:492]: Codacy found an issue: [1 blank line required between summary line and description (found 0) (D205)](https://app.codacy.com/gh/cnescatlab/docker-cat/pullRequest?prid=6253839)

File: tests/test_docker_cat.py [hunk 16]
```diff
@@ -0,0 +1,539 @@
+"""Automated integration test of Docker CAT
```
Comment (ThomasCluzel) [inline tests/test_docker_cat.py:1]: Codacy found an issue: [Multi-line docstring summary should start at the second line (D213)](https://app.codacy.com/gh/cnescatlab/docker-cat/pullRequest?prid=6253839)

File: tests/test_docker_cat.py [hunk 17]
```diff
@@ -0,0 +1,539 @@
...
+        so that it produces a report.
+        """
+        ref = "tests/c_cpp/reference-framac-results.txt"
+        output = "tests/c_cpp/tmp-framac-results.txt"
+        report = "tests/c_cpp/frama-c.csv"
+        cmd = f"frama-c tests/c_cpp/framac/CruiseControl.c tests/c_cpp/framac/CruiseControl_const.c -rte -metrics -report-csv {report}"
+        self.analysis_tool("Frama-C", cmd, ref, output)
+
+    def test_tool_pylint(self):
+        """As a user of this image, I want to run pylint from within a container
```
Comment (ThomasCluzel) [inline tests/test_docker_cat.py:459]: Codacy found an issue: [1 blank line required between summary line and description (found 0) (D205)](https://app.codacy.com/gh/cnescatlab/docker-cat/pullRequest?prid=6253839)

File: tests/test_docker_cat.py [hunk 18]
```diff
@@ -0,0 +1,539 @@
...
+    def teardown_class(cls):
+        """Stop the container
+        """
+        if cls.RUN == "yes":
+            print(f"Stopping {cls.CAT_CONTAINER_NAME}...")
+            docker_client = docker.from_env()
+            docker_client.containers.get(cls.CAT_CONTAINER_NAME).stop()
+
+    def test_up(self):
+        """As a user, I want the server to be UP so that I can use it.
```
Comment (ThomasCluzel) [inline tests/test_docker_cat.py:285]: Codacy found an issue: [One-line docstring should fit on one line with quotes (found 2) (D200)](https://app.codacy.com/gh/cnescatlab/docker-cat/pullRequest?prid=6253839)

File: tests/test_docker_cat.py (hunk 6 above)
Comment (ThomasCluzel) [inline tests/test_docker_cat.py:12]: Codacy found an issue: [Unused import subprocess](https://app.codacy.com/gh/cnescatlab/docker-cat/pullRequest?prid=6253839)

File: tests/test_docker_cat.py [hunk 19]
```diff
@@ -0,0 +1,539 @@
...
+
+    def test_tool_pylint(self):
+        """As a user of this image, I want to run pylint from within a container
+        so that it produces a report.
+        """
+        cmd = "pylint --exit-zero -f json --rcfile=/opt/python/pylintrc_RNC_sonar_2017_A_B tests/python/src/simplecaesar.py"
+        self.analysis_tool("pylint", cmd, "tests/python/reference-pylint-results.json", "tests/python/tmp-pylint-results.json")
+
+    def test_tool_rats(self):
+        """As a user of this image, I want to run RATS from within a container
```
Comment (ThomasCluzel) [inline tests/test_docker_cat.py:466]: Codacy found an issue: [1 blank line required between summary line and description (found 0) (D205)](https://app.codacy.com/gh/cnescatlab/docker-cat/pullRequest?prid=6253839)

File: tests/test_docker_cat.py [hunk 20]
```diff
@@ -0,0 +1,539 @@
...
+        """As a user of this image, I want to run cppcheck from within a container
+        so that it produces a report.
+        """
+        ref = "tests/c_cpp/reference-cppcheck-results.xml"
+        output = "tests/c_cpp/tmp-cppcheck-results.xml"
+        cmd = f"cppcheck --xml-version=2 tests/c_cpp/cppcheck/main.c --output-file={output}"
+        self.analysis_tool("cppcheck", cmd, ref, output, False)
+
+    def test_tool_frama_c(self):
+        """As a user of this image, I want to run Frama-C from within a container
```
Comment (ThomasCluzel) [inline tests/test_docker_cat.py:449]: Codacy found an issue: [1 blank line required between summary line and description (found 0) (D205)](https://app.codacy.com/gh/cnescatlab/docker-cat/pullRequest?prid=6253839)

File: tests/test_docker_cat.py (hunk 14 above)
Comment (ThomasCluzel) [inline tests/test_docker_cat.py:10]: Codacy found an issue: ['platform' imported but unused (F401)](https://app.codacy.com/gh/cnescatlab/docker-cat/pullRequest?prid=6253839)

File: tests/test_docker_cat.py [hunk 21]
```diff
@@ -0,0 +1,539 @@
...
+        """As a user of this image, I want to run RATS from within a container
+        so that it produces a report.
+        """
+        ref = "tests/c_cpp/reference-rats-results.xml"
+        output = "tests/c_cpp/tmp-rats-results.xml"
+        cmd = "rats --quiet --nofooter --xml -w 3 tests/c_cpp/rats"
+        self.analysis_tool("RATS", cmd, ref, output)
+
+    def test_tool_shellcheck(self):
+        """As a user of this image, I want to run shellcheck from within a container
```
Comment (ThomasCluzel) [inline tests/test_docker_cat.py:475]: Codacy found an issue: [1 blank line required between summary line and description (found 0) (D205)](https://app.codacy.com/gh/cnescatlab/docker-cat/pullRequest?prid=6253839)

File: tests/test_docker_cat.py [hunk 22]
```diff
@@ -0,0 +1,539 @@
...
+    RUN = os.environ.get('RUN', "yes")
+    CAT_CONTAINER_NAME = os.environ.get("CAT_CONTAINER_NAME", "cat")
+    CAT_URL = os.environ.get("CAT_URL", "http://localhost:9000")
+    _DOCKER_CAT_IMAGE = "lequal/docker-cat:latest"
+    _PROJECT_ROOT_DIR = str(Path(os.getcwd()).parent)
+
+    # Functions
+    @classmethod
+    def wait_cat_ready(cls, container_name: str):
+        """This function waits for SonarQube to be configured by
```
Comment (ThomasCluzel) [inline tests/test_docker_cat.py:47]: Codacy found an issue: [1 blank line required between summary line and description (found 0) (D205)](https://app.codacy.com/gh/cnescatlab/docker-cat/pullRequest?prid=6253839)

File: tests/test_docker_cat.py [hunk 23]
```diff
@@ -0,0 +1,539 @@
...
+    def test_up(self):
+        """As a user, I want the server to be UP so that I can use it.
+        """
+        status = requests.get(f"{self.CAT_URL}/api/system/status",
+                    auth=("admin", "admin")).json()['status']
+        # Hint: if this test fails, the server might still be starting
+        assert status == "UP"
+
+    def test_check_plugins(self):
+        """As a SonarQube user, I want the plugins listed in the README
```
Comment (ThomasCluzel) [inline tests/test_docker_cat.py:293]: Codacy found an issue: [1 blank line required between summary line and description (found 0) (D205)](https://app.codacy.com/gh/cnescatlab/docker-cat/pullRequest?prid=6253839)

**Comment by ThomasCluzel** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 1
           

Complexity increasing per file
==============================
- tests/test_docker_cat.py  11
         

Clones added
============
- tests/test_docker_cat.py  2
         
```


See the complete overview on [Codacy](https://app.codacy.com/gh/cnescatlab/docker-cat/pullRequest?prid=6253839&bid=20365774)

File: tests/test_docker_cat.py [hunk 24]
```diff
@@ -0,0 +1,566 @@
...
+        so that I can see its level of quality on the SonarQube server.
+        """
+        sensors = (
+            "Sensor C++ (Community) SquidSensor [cxx]",
+            "Sensor SonarFrama-C [framac]"
+        )
+        self.language("C/C++", "c++", "c_cpp", sensors, "c-dummy-project", 0, "CNES_C_A", 1)
+        # 0 issue are expected with the Sonar way Quality Profile for
+        # C++ (Community) because it does not have any rule enabled.
+
```
Comment (ThomasCluzel) [inline tests/test_docker_cat.py:403]: Codacy found an issue: [Trailing whitespace](https://app.codacy.com/gh/cnescatlab/docker-cat/pullRequest?prid=6253839)

**Comment by ThomasCluzel** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- tests/test_docker_cat.py  11
         

Clones added
============
- tests/test_docker_cat.py  2
         
```


See the complete overview on [Codacy](https://app.codacy.com/gh/cnescatlab/docker-cat/pullRequest?prid=6253839&bid=20365774)

**Comment by ThomasCluzel** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- tests/test_docker_cat.py  9
         
```


See the complete overview on [Codacy](https://app.codacy.com/gh/cnescatlab/docker-cat/pullRequest?prid=6253839&bid=20365774)

**Review by begarco — APPROVED** [review]

**Pull request merged**